packages/legend-application-studio/src/stores/GraphEditModeManagerState.ts

```diff
@@ -0,0 +1,1063 @@
+            onError: () =>
+              this.editorStore.applicationStore.setBlockingAlert(undefined),
+          },
+        ),
+      )) as TextCompilationResult;
+      this.editorStore.graphState.setMostRecentCompilationOutcome(
+        GraphCompilationOutcome.SUCCEEDED,
+      );
+
+      this.editorStore.graphState.warnings = compilationResult.warnings
+        ? this.editorStore.graphState.TEMPORARY__removeDependencyProblems(
+            compilationResult.warnings,
+          )
+        : [];
+      this.editorStore.graphState.compilationResultEntities =
+        compilationResult.entities;
+      this.editorStore.applicationStore.setBlockingAlert({
+        message: 'Leaving text mode and rebuilding graph...',
+        showLoading: true,
+      });
+    }
+  }
+
+  *handleCleanupFailure(error: unknown): GeneratorFn<void> {
+    assertErrorThrown(error);
+    this.editorStore.graphState.setMostRecentCompilationOutcome(
+      GraphCompilationOutcome.FAILED,
+    );
+    this.editorStore.graphState.setMostRecentCompilationGraphHash(
+      this.getCurrentGraphHash(),
+    );
+    if (error instanceof EngineError && error.sourceInformation) {
+      this.grammarTextEditorState.setForcedCursorPosition({
+        lineNumber: error.sourceInformation.startLine,
+        column: error.sourceInformation.startColumn,
+      });
+    }
+    this.editorStore.applicationStore.log.error(
+      LogEvent.create(GRAPH_MANAGER_EVENT.COMPILATION_FAILURE),
+      'Compilation failed:',
+      error,
+    );
+    if (this.editorStore.graphManagerState.graphBuildState.hasFailed) {
+      // TODO: when we support showing multiple notification, we can split this into 2 messages
+      this.editorStore.applicationStore.notifyWarning(
+        `Can't build graph, please resolve compilation error before leaving text mode. Compilation failed with error: ${error.message}`,
+      );
+    } else {
+      this.editorStore.applicationStore.notifyWarning(
+        `Compilation failed: ${error.message}`,
+      );
+      this.editorStore.applicationStore.setActionAlertInfo({
+        message: 'Project is not in a compiled state',
+        prompt:
+          'All changes made since the last time the graph was built successfully will be lost',
+        type: ActionAlertType.CAUTION,
+        actions: [
+          {
+            label: 'Discard Changes',
+            default: true,
+            type: ActionAlertActionType.PROCEED_WITH_CAUTION,
+            handler: (): void => {
+              flowResult(
+                this.editorStore.switchModes(GRAPH_EDITOR_MODE.FORM, {
+                  isCompilationFailure: true,
+                }),
+              ).catch(this.editorStore.applicationStore.alertUnhandledError);
+            },
+          },
+          {
+            label: 'Stay',
+            default: true,
+            type: ActionAlertActionType.PROCEED,
+          },
+        ],
+      });
+    }
+  }
+
+  *cleanupBeforeEntering(fallbackOptions?: {
```

Comment:
why is this `fallbackOptions` not being shown as part of the `abstract` class contract?

Reply:
see my comment in `EditorStore`; that is an idea worths exploring here, I think our interface here is still not generic enough.